File: iosApp/ddrhighspeed/views/detail/movie/MovieUtil.swift
import SwiftUI
import shared

class MovieUtil {
    let cyan = Color(UIColor(red: 0, green: 0.79, blue: 0.79, alpha: 1.0))
    let orange = Color(UIColor(red: 0.98, green: 0.63, blue: 0, alpha: 1.0))
    let red = Color(UIColor(red: 1, green: 0, blue: 0, alpha: 0.75))
    let purple = Color(UIColor(red: 0.5, green: 0, blue: 0.5, alpha: 1.0))
    let green = Color(UIColor(red: 0, green: 0.5, blue: 0, alpha: 1.0))
    
    func getLabel(difficulty: String)->(String, Color) {
        switch difficulty {
        case "BESP":
            return ("Single Beginner", cyan)
        case "BSP":
            return ("Single Basic", orange)
        case "DSP":
            return ("Single Difficult", red)
        case "ESP":
            return ("Single Expert", green)
        case "CSP":
            return ("Single Challenge", purple)
        case "BDP":
            return ("Double Basic", orange)
        case "DDP":
            return ("Double Difficult", red)
        case "EDP":
            return ("Double Expert", green)
        case "CDP":
            return ("Double Challenge", purple)
        default:
            return ("Undefined Difficulty", .gray)
        }
    }
    
    let diffOrder: [String] = ["BESP", "BSP", "DSP", "ESP", "CSP", "BDP", "DDP", "EDP", "CDP"]
    func sortByDifficulty(movies: [Movie])-> [Movie] {
        return movies.sorted(by: {
            guard let index1 = diffOrder.firstIndex(of: $0.difficulty)
            else {
                // $0のdifficultyはdiffOrderに含まれないので、$0が後ろになる
                return false
            }
            
            guard let index2 = diffOrder.firstIndex(of: $1.difficulty)
            else {
                // $1のdifficultyはdiffOrderに含まれないので、$0が後ろになる
                return true
            }
            
            return index1 < index2
        })
    }
}

File: shared/src/commonMain/kotlin/com/fias/ddrhighspeed/shared/MovieUtil.kt
package com.fias.ddrhighspeed.shared

import com.fias.ddrhighspeed.shared.cache.Movie

class MovieUtil {
    private val sortOrder = listOf("BESP", "BSP", "DSP", "ESP", "CSP", "BDP", "DDP", "EDP", "CDP")

    private val cyan = ColorRgba(0f, 0.79f, 0.79f, 1.0f)
    private val orange = ColorRgba(0.98f, 0.63f, 0f, 1.0f)
    private val red = ColorRgba(1f, 0f, 0f, 0.75f)
    private val purple = ColorRgba(0.5f, 0f, 0.5f, 1.0f)
    private val green = ColorRgba(0f, 0.5f, 0f, 1.0f)
    private val gray = ColorRgba(0.5f, 0.5f, 0.5f, 1.0f)

    fun createMovieData(movieList: List<Movie>): List<Triple<String, ColorRgba, String>> {
        return sortByDifficulty(movieList)
            .map {
                val (label, colorRgba) = getLabel(it.difficulty)
                Triple(label, colorRgba, it.movie_id)
            }
    }

    private fun getLabel(difficulty: String): Pair<String, ColorRgba> {
        return when (difficulty) {
            "BESP" -> Pair("Single Beginner", cyan)
            "BSP" -> Pair("Single Basic", orange)
            "DSP" -> Pair("Single Difficult", red)
            "ESP" -> Pair("Single Expert", green)
            "CSP" -> Pair("Single Challenge", purple)
            "BDP" -> Pair("Double Basic", orange)
            "DDP" -> Pair("Double Difficult", red)
            "EDP" -> Pair("Double Expert", green)
            "CDP" -> Pair("Double Challenge", purple)
            else -> Pair("Undefined Difficulty", gray)
        }
    }

    private fun sortByDifficulty(movieList: List<Movie>): List<Movie> {
        return movieList.sortedBy { sortOrder.indexOf(it.difficulty) }
    }
}

data class ColorRgba(val red: Float, val green: Float, val blue: Float, val alpha: Float)